src/instruction_graph_generator.cc

```diff
@@ -0,0 +1,2038 @@
+      m_policy(policy), m_memories(m_system.memories.size()) //
+{
+#ifndef NDEBUG
+	assert(m_system.memories.size() <= max_num_memories);
+	assert(std::all_of(m_system.devices.begin(), m_system.devices.end(),
+	    [&](const instruction_graph_generator::device_info& device) { return device.native_memory < m_system.memories.size(); }));
+	for(memory_id mid_a = 0; mid_a < m_system.memories.size(); ++mid_a) {
+		assert(m_system.memories[mid_a].copy_peers[mid_a]);
+		for(memory_id mid_b = mid_a + 1; mid_b < m_system.memories.size(); ++mid_b) {
+			assert(m_system.memories[mid_a].copy_peers[mid_b] == m_system.memories[mid_b].copy_peers[mid_a]
+			       && "system_info::memories::copy_peers must be reflexive");
+		}
+	}
+#endif
+
+	batch init_epoch_batch;
+	m_idag->begin_epoch(task_manager::initial_epoch_task);
+	const auto init_epoch = create<epoch_instruction>(init_epoch_batch, task_manager::initial_epoch_task, epoch_action::none, instruction_garbage{},
+	    [](const auto& record_debug_info) { record_debug_info(command_id(0 /* or so we assume */)); });
+	m_last_epoch = init_epoch;
+	flush_batch(std::move(init_epoch_batch));
+
+	// The root collective group (aka MPI_COMM_WORLD) already exists, but we must still equip it with a meaningful last_host_task.
+	m_collective_groups.emplace(root_collective_group_id, init_epoch);
+}
+
+void generator_impl::notify_buffer_created(
+    const buffer_id bid, const range<3>& range, const size_t elem_size, const size_t elem_align, allocation_id user_aid) //
+{
+	const auto [iter, inserted] =
+	    m_buffers.emplace(std::piecewise_construct, std::tuple(bid), std::tuple(range, elem_size, elem_align, m_system.memories.size()));
+	assert(inserted);
+
+	if(user_aid != null_allocation_id) {
+		// The buffer was host-initialized through a user-specified pointer, which we consider a fully coherent allocation in user_memory_id.
+		assert(user_aid.get_memory_id() == user_memory_id);
+		const box entire_buffer = subrange({}, range);
+
+		auto& buffer = iter->second;
+		auto& memory = buffer.memories[user_memory_id];
+		auto& allocation = memory.allocations.emplace_back(user_aid, nullptr /* alloc_instruction */, entire_buffer, buffer.range);
+
+		allocation.track_atomic_write(entire_buffer, m_last_epoch);
+		buffer.track_original_write(entire_buffer, m_last_epoch, user_memory_id);
+	}
+}
+
+void generator_impl::notify_buffer_debug_name_changed(const buffer_id bid, const std::string& name) { m_buffers.at(bid).debug_name = name; }
+
+void generator_impl::notify_buffer_destroyed(const buffer_id bid) {
+	const auto iter = m_buffers.find(bid);
+	assert(iter != m_buffers.end());
+	auto& buffer = iter->second;
+
+	batch free_batch;
+	for(memory_id mid = 0; mid < buffer.memories.size(); ++mid) {
+		auto& memory = buffer.memories[mid];
+		if(mid == user_memory_id) {
+			for(const auto& user_alloc : memory.allocations) {
```

Comment:
Yes - the user-pointer is considered to be backing allocation that lives as long as the buffer. Fences are not treated as buffer allocations.

Reply:
Ok but then there could only ever be one allocation here, right? (I.e., should this be a loop?)

Reply:
Yes, either 0 or 1. I would argue that a loop is the natural construct here because the data structure is a vector (because it's used for all memories, not just user memory). I've extended the comment though saying that it should hold 0 or 1 element in practice.